### DISCUSSION TITLE
How to render object on cylinder

### DISCUSSION THREAD

**Discussion by L2zz** [sceneview/sceneview#167] [Q&A]
Hi! 
I'm going to implement an application that allows an object (Andy) to rotate along the side of the cylinder.
I'd like some advice on this. Specifically, the conditions and questions are as follows.

### Overview of the target App
<p align="center">
<img src="https://user-images.githubusercontent.com/37107066/190966370-dd229bcd-74af-4449-abac-07023ff8d96f.png" width="20%"/>
</p>

### Conditions
**Cylinder**
- Rendering using geospatial API
- Scalable radius and height

**Andy**
- Rendering on the side of the cylinder
- Rotatable along the side of the cylinder

### Questions
**1. How to scale the radius or height of the cylinder?**
As far as I know, `ArModelNode` is the only one that supports geospatial API.
The `ArModelNode` can be scaled through a `scaleToUnits`, but this changes the overall size.
What I want is to change the radius or height separately according to the user's request.
So, I wonder **how to specify the scales for the x, y, and z axes individually for the model.**

**2. How to rotate Andy along the side of the cylinder?**
I think directly rendering the Andy model along the side of the cylinder model seems difficult.
So I divide this process into two steps.
1. Rendering Andy as away as the radius from the centerline of the cylinder.
2. Rotate Andy relative to the centerline of the cylinder.

For the above steps, I would like to ask for a way to **render and rotate the model**  
**relative to the axis outside the object (here, the centerline of the cylinder).**

Thanks very much for supporting this wonderful library 👍.

**Comment by ThomasGorisse** [comment]
Hi,

Very interesting and clear question (even if you should better put it in the [Discussion](https://github.com/SceneView/sceneview-android/discussions) section).

What I would suggest is having an `ArModelNode` with `Plane` placement that displays a cylinder coming from a glb for now.
For now because it will soon (v1.0.0) be easier and more efficient to use a `Cylinder` `Geometry` (It will look like this: https://gist.github.com/ThomasGorisse/7e07681e6aae794982ced3bc61af7f8d)
But if it is well defined inside Blender, it's not much more consuming to have it coming from a glTF.

You can then change its dimensions using the 
https://github.com/SceneView/sceneview-android/blob/b7341cced9c0d7ac036d3a039a5c5fa9a8d39c1a/sceneview/src/main/java/io/github/sceneview/node/Node.kt#L184
Keep `z` at `1.0f` and adjust the radius with `scale.x` and the height with `scale.y`
![image](https://user-images.githubusercontent.com/6597529/190991723-8a2199e3-f910-48c1-987f-164e21cd8edb.png)

For the Andy part, depending on if you want your Cylinder to rotate with him or not, you can simply add it has a child or as sub-child `ModelNode`:
https://github.com/SceneView/sceneview-android/blob/b7341cced9c0d7ac036d3a039a5c5fa9a8d39c1a/sceneview/src/main/java/io/github/sceneview/node/NodeParent.kt#L44
If you want it to be independent from the Cylinder, add the Cylinder as a child of your `ArNode` instead of loading the glb at the ArNode level.

For the rotation part, all you have to do is to put it at a certain distance using 
https://github.com/SceneView/sceneview-android/blob/b7341cced9c0d7ac036d3a039a5c5fa9a8d39c1a/sceneview/src/main/java/io/github/sceneview/node/Node.kt#L143
from the pivot `ArNode` and then rotate the `ArNode`.

You can also have a look at the https://github.com/SceneView/sceneview-android/blob/b7341cced9c0d7ac036d3a039a5c5fa9a8d39c1a/sceneview/src/main/java/io/github/sceneview/node/Node.kt#L734
:-) function for controlling what is Andy looking at during rotation.

Pay attention to the fact that depending on your `ArNode.placementNode` the rotation of your `ArNode` might be influenced by ARCore positionning.
For that reason you should have something like this:

```
ArModelNode
    > CylinderNode
    > PivotNode
       > AndyNode
```

**Comment by L2zz** [comment]
Thanks for your great help! And I'm sorry for using the wrong tab😂.

I have some more questions.

> For the rotation part, all you have to do is to put it at a certain distance using
> ``` 
> var position: Position = position
> ```
> from the pivot ArNode and then rotate the ArNode.

The code seems to adjust the position by directly entering the value.
I didn't specify it above, but I want is rotating Andy through the user's gesture,
so the question is,
 **Q. How to fix the distance (here, radius) between the pivot and Andy during interaction with the user?**

And I have a simple question about the relationship between the child and parent node.
**Q. If I modify the scale of the parent, does the unit of the scale and position of the child are affected?**

**Comment by L2zz** [comment]
I solved it in the way you suggested before. Thx!
